CertificateOwnsSecret panics with index out of range [0] when Certificate is deleted during namespace teardown

Describe the bug:

  cert-manager-controller panics with index out of range [0] with length 0 in CertificateOwnsSecret() at internal/controller/certificates/certificates.go:94.

  The function accesses ownerReferences[0] unconditionally, but a Secret can have zero ownerReferences if the Certificate was deleted (e.g., due to namespace deletion)
  while the trigger controller was mid-reconcile. When the list is empty, the index access panics and the controller crashes with exit code 2.

  panic: runtime error: index out of range [0] with length 0

  goroutine 491 [running]:
  github.com/cert-manager/cert-manager/internal/controller/certificates.CertificateOwnsSecret(...)
        internal/controller/certificates/certificates.go:94 +0x545
  github.com/cert-manager/cert-manager/pkg/controller/certificates/trigger.(*controller).ProcessItem(...)
        pkg/controller/certificates/trigger/trigger_controller.go:176 +0x3f4

  Expected behaviour:

  CertificateOwnsSecret() should guard against an empty ownerReferences slice before accessing index [0], and return false (or an appropriate value) rather than panicking.
  The trigger controller should handle a missing/deleted Certificate gracefully without crashing.

  Steps to reproduce the bug:

  This is a narrow race condition requiring all three to occur simultaneously:

  1. A cert-manager reconcile is actively running for a specific Certificate
  2. That Certificate's namespace begins deletion while the reconcile is in progress
  3. The Certificate is evicted from cert-manager's internal cache between the start of ProcessItem() and the ownerReferences[0] access in CertificateOwnsSecret()

  In our case:
  - A Certificate (953efcef-...-auth-proxy-keypair) was flagged for re-issuance because its Secret did not exist (reason: DoesNotExist)
  - The trigger controller called CertificateOwnsSecret() to verify Secret ownership
  - The Secret had zero ownerReferences at that point → panic

  Anything else we need to know?:

  - The pod had been running for 176 days without issue before this occurred — confirming the narrow timing window
  - It has crashed exactly once in our environment
  - The controller self-recovered immediately on restart with no data loss
  - No alerts fired prior to the crash; it was caught via log monitoring
  - Suggested fix: add a bounds check before ownerReferences[0] in CertificateOwnsSecret():
  if len(ownerReferences) == 0 {
      return false
  }

  Environment details:
  - Kubernetes version: EKS-A
  - Cloud-provider/provisioner: Bare-metal / on-prem
  - cert-manager version: v1.16.4-eks-a-97
  - Install method: Helm
  - error is coming from this line :- https://github.com/cert-manager/cert-manager/blob/c928c6e8ee68eff657bf56e0f8e11999ebe9f233/internal/controller/certificates/certificates.go#L94

  /kind bug

got the same panic:
```
panic: runtime error: index out of range [0] with length 0 [recovered, repanicked]

goroutine 523 [running]:
sync.(*WaitGroup).Go.func1.1()
        sync/waitgroup.go:251 +0x45
panic({0x2e74060?, 0xc7c49718630?})
        runtime/panic.go:860 +0x13a
github.com/cert-manager/cert-manager/internal/controller/certificates.CertificateOwnsSecret({0x3248f48?, 0xc7c4a75ea80?}, {0x322e1b0?, 0xc7c4387ca00?}, {0x3219860, 0xc7c43c72780}, 0xc7c4fdcf208)
        github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/internal/controller/certificates/certificates.go:94 +0x5c5
github.com/cert-manager/cert-manager/pkg/controller/certificates/trigger.(*controller).ProcessItem(0xc7c43c68780, {0x3248f48, 0xc7c441b0af0}, {{0xc7c4efec8c0, 0xc}, {0xc7c4a75ea80, 0x16}})
        github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/certificates/trigger/trigger_controller.go:180 +0x40d
github.com/cert-manager/cert-manager/pkg/controller.(*controller).worker.func1(0xc7c43c6a480, {{0xc7c4efec8c0?, 0xf?}, {0xc7c4a75ea80?, 0x0?}}, {{0x3254b00?, 0xc7c4331c880?}, 0xc7c4b281dd0?}, {0x3248f48, 0xc7c441b0af0})
        github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/controller.go:147 +0x170
github.com/cert-manager/cert-manager/pkg/controller.(*controller).worker(0xc7c43c6a480, {0x3248f48, 0xc7c441b0af0})
        github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/controller.go:164 +0xc5
github.com/cert-manager/cert-manager/pkg/controller.(*controller).Run.func1()
        github.com/cert-manager/cert-manager@v0.0.0-00010101000000-000000000000/pkg/controller/controller.go:107 +0x25
sync.(*WaitGroup).Go.func1()
        sync/waitgroup.go:258 +0x4a
created by sync.(*WaitGroup).Go in goroutine 245
        sync/waitgroup.go:238 +0x73
```

cert-manager version is 1.20.2

--- a/internal/controller/certificates/certificates.go
+++ b/internal/controller/certificates/certificates.go
@@ -64,6 +64,15 @@ func CertificateOwnsSecret(
 		}
 	}
 
+	// If the passed Certificate is not present in the lister's cache (for
+	// example, because it was deleted while a reconcile was in progress, such
+	// as during namespace teardown), duplicateCrts will be empty. Return false
+	// so the caller treats the Certificate as not owning the Secret rather
+	// than panicking when indexing the slice below.
+	if len(duplicateCrts) == 0 {
+		return false, nil, nil
+	}
+
 	// If there are no duplicates, return early.
 	if len(duplicateCrts) == 1 && duplicateCrts[0].Name == crt.Name {
 		return true, nil, nil

--- a/internal/controller/certificates/certificates_test.go
+++ b/internal/controller/certificates/certificates_test.go
@@ -56,9 +56,10 @@ func TestCertificateOwnsSecret(t *testing.T) {
 		secrets             []runtime.Object
 		certificates        []runtime.Object
 
-		expectedResult      bool
-		expectedOtherOwners []string
-		expectedError       error
+		expectedResult            bool
+		expectedOtherOwners       []string
+		expectedError             error
+		ignoreSelectedCertificate bool
 	}{
 		{
 			name: "Certificate is only cert referencing the secret",
@@ -129,6 +130,18 @@ func TestCertificateOwnsSecret(t *testing.T) {
 			expectedError:       nil,
 		},
 		{
+			name: "Certificate is not present in the lister cache (e.g. deleted mid-reconcile)",
+
+			selectedCertificate: "certificate-missing",
+			secrets:             []runtime.Object{},
+			certificates:        []runtime.Object{},
+
+			expectedResult:            false,
+			expectedOtherOwners:       nil,
+			expectedError:             nil,
+			ignoreSelectedCertificate: true,
+		},
+		{
 			name: "Certificate has conflict, is the oldest, but annotation marks another as the owner",
 
 			selectedCertificate: "certificate-3",
@@ -188,7 +201,10 @@ func TestCertificateOwnsSecret(t *testing.T) {
 				}
 			}
 			if selectedCrt == nil {
-				t.Fatal("failed to find selected Certificate")
+				if !tt.ignoreSelectedCertificate {
+					t.Fatal("failed to find selected Certificate")
+				}
+				selectedCrt = certificate(tt.selectedCertificate, testCreationTimestamp)
 			}
 
 			// Call the function under test

